ts/kpt-functions/src/testing.ts

```diff
@@ -129,12 +129,13 @@ async function testFn(
   }
 }
 
-function valueOf(configs: Configs): string {
+function valueOf(configs: Configs): any {
   const output = configs.toResourceList();
-  return safeDump(output, {
+  const yaml = safeDump(output, {
     indent: 2,
     noArrayIndent: true,
     skipInvalid: true,
     sortKeys: true,
   });
+  return safeLoad(yaml);
```

Comment:
I know that it's a roundabout way to implement it. But in practice, returning `configs.toResourceList()` causes many test failures for us. Mainly these two types of inequalities:
```
Expected $.items[0].metadata not to have properties
        namespace: undefined
```
and
```
Expected $.items[0] to be a kind of IAMPolicyResource, but was Object({ apiVersion: 'iam.cnrm.cloud.google.com/v1alpha2', kind: 'IAMPolicy', metadata: Object({ name: 'iam-policy', namespace: 'project-level' }), spec: Object({ resourceRef: Object({ apiVersion: 'test', kind: 'Test' }), bindings: [  ] }) }).
```

The first kind of inequality is easy to understand. `safeDump` elides all `undefined` properties.

The second kind of inequality results from the call to `deepCopy` on line 102 https://github.com/GoogleContainerTools/kpt-functions-sdk/blob/cfd762890f962f854cda1c8cb424a9fd4195d6f5/ts/kpt-functions/src/testing.ts#L102
I haven't dug into the implementation of `Configs.deepCopy`. But basically, before the equality comparison happens, the `input` object goes through a `deepCopy` and the `expectedOutput` object doesn't. That `deepCopy` is apparently not "equality-safe", however whatever differences it creates disappear again upon `safeDump`.

We have already relied on the semantics of post-yaml comparison in our test cases. 44/118 of our test cases break if we directly return `configs.toResourceList()` from `valueOf`. `safeDump` returns a string. And if strings `s` and `t` are equal (and valid YAML), then `safeLoad(s)` and `safeLoad(t)` must also be equal (`safeDump` is deterministic, surely). The proposed change can't break existing tests.

## Alternatives
### Patch the differences
I suppose it's possible to address the two types of equality differences that we have observed above. We can do a `deepCopy` of the `o` object that goes into the comparison. We can explicitly filter out `undefined` properties with some recursive function. Or maybe modify our kpt functions to be meticulous about this, so that all the `undefined`s match up.

But how can we know that these two equality differences are the only ones that could appear? Someone else might depend on other aspects of `safeDump` in their tests. It seems a bit arbitrary to patch over these two things specifically.

### Rich YAML comparison
Maybe the `expect(...).toEqual(...)` can be replaced with some YAML-specific equality comparison function, which takes two YAMLs and outputs a rich diff on the strings? For example a line-based diff. Then we can go back to only using `safeDump` and not `safeLoad`. That should _probably_ also output readable diffs.

Reply:
Thanks for the detailed explanation.